File: analytics.js
const dayMs = 86400000;
export const toIso = date => new Date(date.getTime() - date.getTimezoneOffset() * 60000).toISOString().slice(0, 10);
export const fromIso = value => new Date(`${value}T12:00:00`);
export function startOfWeek(date) { const d = new Date(date); const day = d.getDay() || 7; d.setDate(d.getDate() - day + 1); d.setHours(12,0,0,0); return d; }
export const addDays = (date, days) => new Date(date.getTime() + days * dayMs);
export function periodBounds(type, reference) {
  const d = new Date(reference); d.setHours(12,0,0,0);
  if (type === 'week') { const start = startOfWeek(d); return [start, addDays(start, 6)]; }
  if (type === 'year') return [new Date(d.getFullYear(),0,1,12), new Date(d.getFullYear(),11,31,12)];
  return [new Date(d.getFullYear(),d.getMonth(),1,12), new Date(d.getFullYear(),d.getMonth()+1,0,12)];
}
export function buildReport(data, type, reference) {
  const [start,end] = periodBounds(type, reference); const min=toIso(start), max=toIso(end);
  const byOperator = data.operators.map(operator => {
    const counts = Object.fromEntries(data.activities.map(a => [a.id,0])); let total=0; let activeDays=0;
    Object.entries(data.assignments).forEach(([key,ids]) => {
      const [op,date] = key.split('|'); if(op!==operator.id || date<min || date>max || !Array.isArray(ids)) return;
      if(ids.length) activeDays++; ids.forEach(id => { if(id in counts){counts[id]++;total++;} });
    });
    return { operator, counts, total, activeDays, percentages:Object.fromEntries(Object.entries(counts).map(([id,n])=>[id,total?n/total:0])) };
  });
  return { start, end, byOperator, totalAssignments:byOperator.reduce((n,o)=>n+o.total,0), activeOperators:byOperator.filter(o=>o.total).length };
}

File: lib/analytics.dart
import 'models.dart';

enum PeriodType { week, month, year }

DateTime startOfWeek(DateTime d) =>
    DateTime(d.year, d.month, d.day).subtract(Duration(days: d.weekday - 1));
(DateTime, DateTime) periodBounds(PeriodType type, DateTime reference) {
  if (type == PeriodType.week) {
    final start = startOfWeek(reference);
    return (start, start.add(const Duration(days: 6)));
  }
  if (type == PeriodType.year) {
    return (DateTime(reference.year), DateTime(reference.year, 12, 31));
  }
  return (
    DateTime(reference.year, reference.month),
    DateTime(reference.year, reference.month + 1, 0),
  );
}

class OperatorReport {
  const OperatorReport({
    required this.operator,
    required this.counts,
    required this.total,
    required this.activeDays,
  });
  final OperatorModel operator;
  final Map<String, int> counts;
  final int total;
  final int activeDays;
  double percent(String activityId) =>
      total == 0 ? 0 : (counts[activityId] ?? 0) / total;
}

List<OperatorReport> buildReport(
  AppData data,
  PeriodType type,
  DateTime reference,
) {
  final (start, end) = periodBounds(type, reference);
  final min = isoDate(start), max = isoDate(end);
  return data.operators.map((operator) {
    final counts = {for (final a in data.activities) a.id: 0};
    var activeDays = 0;
    for (final entry in data.assignments.entries) {
      final parts = entry.key.split('|');
      if (parts.length != 2 ||
          parts[0] != operator.id ||
          parts[1].compareTo(min) < 0 ||
          parts[1].compareTo(max) > 0) {
        continue;
      }
      if (entry.value.isNotEmpty) {
        activeDays++;
      }
      for (final id in entry.value) {
        if (counts.containsKey(id)) counts[id] = counts[id]! + 1;
      }
    }
    return OperatorReport(
      operator: operator,
      counts: counts,
      total: counts.values.fold(0, (a, b) => a + b),
      activeDays: activeDays,
    );
  }).toList();
}

